livemaker/archive.py

```diff
@@ -767,6 +767,30 @@ def write(self, filename, arcname=None, compress_type=None, unk1=None):
             FileExistsError: If an entry matching `arcname` already exists in this archive.
             UnsupportedLiveMakerCompression: If the `compress_type` is unsupported.
 
+        """
+        write_as(self, filename, filename, arcname, compress_type, unk1)
+
+    def write_as(self, filename, packname, arcname=None, compress_type=None, unk1=None):
+        """Write the file `filename`named as `packname` into the archive.
+
+        Args:
+            filename: File to write into archive.
+            packname: Filename to be written into archive.
+            arcname: If given, the archive file entry will be named `arcname`.
+                By default, `arcname` will be the same as `filename`,
+                but with any drive letter and leading path separators removed.
+                Posix paths will be replaced with equivalent Windows paths.
```

Comment:
The original function uses "filename" for both filesystem and archive. This requires the files and subdirectories in the current working directory. This also does not support absolute filesystem paths (and not to talk about things like "~" or ".."). 
The new function just separates archive name and filesystem name. This enables us to use absolute path names, without the need to strip the archive path apart.
The reason for implement the new parameter as a new function is just a habit. I do really few in python, so I was not aware of the pythons optional parameter thing.

Reply:
The original function was supposed to have that separated file system and archive name behavior, although it looks like there was a bug which you ended up fixing:

```diff
-        with open(name, 'rb') as f:
+        with open(filename, 'rb') as f:
```

The original function with that change does the same thing as the new `write_as()` function

Reply:
right

Reply:
Ahmm. Nope.
The original function was this:
    def write(self, filename, arcname=None, compress_type=None, unk1=None):
filename for both archive and filesystem name.
The write_as I send as PR hads a mistake:
livemaker/archive.py line 800 should be
arcpath = PureWindowsPath(packname)
So this function does not the same anymore

Reply:
the original function has

```
        if arcname is None:
            arcpath = PureWindowsPath(filename)
        else:
            arcpath = PureWindowsPath(arcname)
```

so using `write(filename, arcname=packname, ...)` is the same as your `write_as(filename, packname)`.

Reply:
basically I would really like to avoid adding a new `write_as` function, since the original `write()` function is intentionally mirroring the API used for other standard python libs (`zipfile`, `tarfile`), where:

* `filename` should always be any (absolute or relative) path to some file on the user's machine
* if `arcname=...` is provided and not None, it should be the relative path to the location inside the zip/tar/livemaker archive

Reply:
Ah, finally got it. I had taken arcname as the archive (.exe or .ext) file name...

Reply:
ah, yeah in this case the argument is just called `arcname` because that's how it is named in other standard python libs: https://docs.python.org/3.7/library/zipfile.html#zipfile.ZipFile.write